// @progress/kendo-react-spreadsheet/messages.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
const e = "spreadsheet.bold", t = "spreadsheet.italic", s = "spreadsheet.underline", o = "spreadsheet.openFile", n = "spreadsheet.textColor", a = "spreadsheet.background", r = "spreadsheet.addRowAbove", d = "spreadsheet.addRowBelow", l = "spreadsheet.addColumnLeft", i = "spreadsheet.addColumnRight", c = "spreadsheet.deleteRow", p = "spreadsheet.deleteColumn", h = "spreadsheet.undo", g = "spreadsheet.redo", m = "spreadsheet.fontSize", f = "spreadsheet.fontSizeIncrease", u = "spreadsheet.fontSizeDecrease", x = "spreadsheet.fontName", y = "spreadsheet.cleanFormatting", A = "spreadsheet.format", w = "spreadsheet.alignHorizontally", C = "spreadsheet.alignVertically", b = "spreadsheet.align", S = "spreadsheet.alignLeft", z = "spreadsheet.alignRight", D = "spreadsheet.alignCenter", R = "spreadsheet.center", T = "spreadsheet.alignJustify", B = "spreadsheet.justify", F = "spreadsheet.alignTop", N = "spreadsheet.alignMiddle", k = "spreadsheet.alignBottom", I = "spreadsheet.increaseDecimal", L = "spreadsheet.decreaseDecimal", M = "spreadsheet.textWrap", E = "spreadsheet.gridLines", U = "spreadsheet.exportToExcel", v = "spreadsheet.nameBox", j = "spreadsheet.defaultSheetName", H = "spreadsheet.home", J = "spreadsheet.file", W = "spreadsheet.insert", O = "spreadsheet.formatTab", P = "spreadsheet.addNewSheet", V = "spreadsheet.error", K = "spreadsheet.ok", q = "spreadsheet.modifyMerged", G = "spreadsheet.cannotModifyDisabled", Q = "spreadsheet.openUnsupported", X = "spreadsheet.duplicateSheetName", Y = {
  [e]: "Bold",
  [t]: "Italic",
  [s]: "Underline",
  [h]: "Undo",
  [g]: "Redo",
  [m]: "Font size",
  [f]: "Increase Font size",
  [u]: "Decrease Font size",
  [x]: "Font",
  [y]: "Clean formatting",
  [A]: "Custom format...",
  [w]: "Align horizontally",
  [C]: "Align vertically",
  [b]: "Align",
  [S]: "Align left",
  [z]: "Align right",
  [D]: "Align center",
  [T]: "Align justify",
  [R]: "Center",
  [B]: "Justify",
  [F]: "Align top",
  [N]: "Align middle",
  [k]: "Align bottom",
  [I]: "Increase decimal",
  [L]: "Decrease decimal",
  [n]: "Text Color",
  [a]: "Background color",
  [r]: "Add row above",
  [d]: "Add row below",
  [l]: "Add column left",
  [i]: "Add column right",
  [c]: "Delete row",
  [p]: "Delete column",
  [o]: "Open...",
  [U]: "Export to Excel",
  [M]: "Text Wrap",
  [E]: "Toggle gridlines",
  [v]: "Name Box",
  [j]: "Sheet",
  [H]: "Home",
  [J]: "File",
  [W]: "Insert",
  [P]: "Add new sheet",
  [O]: "Format",
  [q]: "Cannot change part of a merged cell.",
  [G]: "Cannot modify disabled cells.",
  [Q]: "Unsupported format. Please select an .xlsx file.",
  [X]: "There is an existing sheet with this name. Please enter another name.",
  [V]: "Error",
  [K]: "OK"
}, Z = {
  bold: e,
  italic: t,
  underline: s,
  undo: h,
  redo: g,
  fontSize: m,
  fontSizeIncrease: f,
  fontSizeDecrease: u,
  fontName: x,
  cleanFormatting: y,
  format: A,
  alignHorizontally: w,
  alignVertically: C,
  align: b,
  alignLeft: S,
  alignRight: z,
  alignCenter: D,
  center: R,
  alignJustify: T,
  justify: B,
  alignTop: F,
  alignMiddle: N,
  alignBottom: k,
  increaseDecimal: I,
  decreaseDecimal: L,
  textColor: n,
  background: a,
  addRowAbove: r,
  addRowBelow: d,
  addColumnLeft: l,
  addColumnRight: i,
  deleteRow: c,
  deleteColumn: p,
  open: o,
  exportToExcel: U,
  textWrap: M,
  gridLines: E,
  nameBox: v,
  defaultSheetName: j,
  file: J,
  home: H,
  insert: W,
  addNewSheet: P,
  error: V,
  ok: K,
  modifyMerged: q,
  cannotModifyDisabled: G,
  openUnsupported: Q,
  duplicateSheetName: X,
  formatTab: O
};
export {
  Z as keys,
  Y as messages
};
