// @progress/kendo-react-treelist/rows/TreeListRow.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as t from "react";
import { classNames as x } from "@progress/kendo-react-common";
import { TREELIST_ROW_INDEX_ATTRIBUTE as k } from "../constants/index.mjs";
const u = t.forwardRef((e, o) => {
  const { isSelected: i, rowHeight: a, isAltRow: r, expanded: l, ariaRowIndex: d, level: c, ariaSetSize: s, ariaPosInSet: m } = e, w = x("k-table-row", "k-master-row", {
    "k-selected": i,
    "k-alt": r
  }), R = t.useMemo(
    () => a !== void 0 ? { height: a + "px" } : void 0,
    [a]
  ), n = /* @__PURE__ */ t.createElement(
    "tr",
    {
      onClick: e.onClick,
      onDoubleClick: e.onDoubleClick,
      onContextMenu: e.onContextMenu,
      className: w,
      style: R,
      ref: o,
      "aria-expanded": l,
      role: "row",
      "aria-rowindex": d,
      "aria-level": c.length,
      "aria-posinset": m,
      "aria-setsize": s,
      [k]: e.rowIndex
    },
    e.children
  );
  return /* @__PURE__ */ t.createElement(t.Fragment, null, e.render ? e.render.call(void 0, n, e) : n);
});
u.displayName = "KendoReactTreeListRow";
export {
  u as TreeListRow
};
