// @progress/kendo-react-pivotgrid/messages/index.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
const o = "pivotgrid.emptyCellAriaLabel", t = "pivotgrid.fieldMenuReset", i = "pivotgrid.fieldMenuApply", r = "pivotgrid.configuratorTitle", e = "pivotgrid.configuratorCancel", n = "pivotgrid.configuratorApply", c = "pivotgrid.configuratorColumnsTitle", s = "pivotgrid.configuratorRowsTitle", a = "pivotgrid.configuratorMeasuresTitle", l = "pivotgrid.configuratorEditorSearchTitle", g = "pivotgrid.configuratorEditorSearchPlaceholder", u = "pivotgrid.configuratorMissingFields", p = "pivotgrid.configuratorButtonLabel", d = {
  [o]: "PivotGrid component. Use the arrow keys to navigate.",
  [t]: "Reset",
  [i]: "Apply",
  [e]: "Cancel",
  [n]: "Apply",
  [r]: "Settings",
  [l]: "Fields",
  [g]: "Search...",
  [c]: "Columns",
  [s]: "Rows",
  [a]: "Values",
  [p]: "Change settings",
  [u]: "Select some fields to begin setup"
};
export {
  n as configuratorApply,
  p as configuratorButtonLabel,
  e as configuratorCancel,
  c as configuratorColumnsTitle,
  g as configuratorEditorSearchPlaceholder,
  l as configuratorEditorSearchTitle,
  a as configuratorMeasuresTitle,
  u as configuratorMissingFields,
  s as configuratorRowsTitle,
  r as configuratorTitle,
  o as emptyCellAriaLabel,
  i as fieldMenuApply,
  t as fieldMenuReset,
  d as messages
};
