// @progress/kendo-react-gantt/constants/index.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import { MS_PER_DAY as F, MS_PER_HOUR as U, MS_PER_MINUTE as m } from "@progress/kendo-date-math";
import { TABLE_ROW_INDEX_ATTRIBUTE as T, TABLE_COL_INDEX_ATTRIBUTE as _, TABLE_PREVENT_SELECTION_ELEMENT as t } from "@progress/kendo-react-data-tools";
const n = T, N = _, o = t, s = (/* @__PURE__ */ new Date()).getTime(), I = 7, d = "hh:mm a", R = { skeleton: "MEd" }, c = "MMM", i = "yyyy", L = 100, O = 6, E = "data-dependency-drag-handle", A = "start", D = "finish", M = { [E]: A }, S = { [E]: D }, r = "data-task-id", a = {
  id: "id",
  start: "start",
  end: "end",
  title: "title",
  percentComplete: "percentComplete",
  parentId: "parentId",
  isRollup: "isRollup",
  isExpanded: "isExpanded",
  isInEdit: "isInEdit",
  children: "children",
  isSelected: "isSelected"
}, p = {
  id: "id",
  fromId: "fromId",
  toId: "toId",
  type: "type"
};
export {
  I as DAYS_IN_WEEK_COUNT,
  R as DAY_DATE_FORMAT,
  L as DEFAULT_COLUMN_WIDTH,
  s as DEFAULT_DATE,
  p as DEFAULT_DEPENDENCY_MODEL_FIELDS,
  a as DEFAULT_TASK_MODEL_FIELDS,
  E as DEPENDENCY_DRAG_HANDLE,
  S as DEPENDENCY_DRAG_HANDLE_FINISH_ATT,
  M as DEPENDENCY_DRAG_HANDLE_START_ATT,
  D as DRAG_DIRECTION_FINISH,
  A as DRAG_DIRECTION_START,
  N as GANTT_COL_INDEX_ATTRIBUTE,
  o as GANTT_PREVENT_SELECTION_ELEMENT,
  n as GANTT_ROW_INDEX_ATTRIBUTE,
  d as HOUR_DATE_FORMAT,
  O as MILESTONE_OFFSET,
  c as MONTH_DATE_FORMAT,
  F as MS_PER_DAY,
  U as MS_PER_HOUR,
  m as MS_PER_MINUTE,
  r as TASK_ID_ATT,
  i as YEAR_DATE_FORMAT
};
