// @progress/kendo-react-gantt/editors/FormInput.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as e from "react";
import { FieldWrapper as b } from "@progress/kendo-react-form";
import { Label as u, Hint as I, Error as $ } from "@progress/kendo-react-labels";
import { Input as v } from "@progress/kendo-react-inputs";
const y = (c) => {
  const { validationMessage: o, touched: m, label: p, id: t, valid: a, disabled: i, hint: n, type: f, optional: h, ...E } = c, r = m && o, l = !r && n, s = l ? `${t}_hint` : "", d = r ? `${t}_error` : "";
  return /* @__PURE__ */ e.createElement(b, { className: "k-col-span-full" }, /* @__PURE__ */ e.createElement(u, { editorId: t, editorValid: a, editorDisabled: i, optional: h }, p), /* @__PURE__ */ e.createElement("div", { className: "k-form-field-wrap" }, /* @__PURE__ */ e.createElement(
    v,
    {
      valid: a,
      type: f,
      id: t,
      disabled: i,
      ariaDescribedBy: `${s} ${d}`,
      ...E
    }
  ), l && /* @__PURE__ */ e.createElement(I, { id: s }, n), r && /* @__PURE__ */ e.createElement($, { id: d }, o)));
};
export {
  y as FormInput
};
