// @progress/kendo-react-gantt/editors/FormNumericTextBox.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as e from "react";
import { FieldWrapper as E } from "@progress/kendo-react-form";
import { Label as h, Hint as b, Error as x } from "@progress/kendo-react-labels";
import { NumericTextBox as u } from "@progress/kendo-react-inputs";
const B = (m) => {
  const { validationMessage: i, visited: c, label: p, id: t, valid: o, disabled: a, hint: d, ...f } = m, r = c && i, n = !r && d, s = n ? `${t}_hint` : "", l = r ? `${t}_error` : "";
  return /* @__PURE__ */ e.createElement(E, null, /* @__PURE__ */ e.createElement(h, { editorId: t, editorValid: o, editorDisabled: a }, p), /* @__PURE__ */ e.createElement("div", { className: "k-form-field-wrap" }, /* @__PURE__ */ e.createElement(
    u,
    {
      ariaDescribedBy: `${s} ${l}`,
      valid: o,
      id: t,
      disabled: a,
      format: "p0",
      step: 0.01,
      ...f
    }
  ), n && /* @__PURE__ */ e.createElement(b, { id: s }, d), r && /* @__PURE__ */ e.createElement(x, { id: l }, i)));
};
export {
  B as FormNumericTextBox
};
