// @progress/kendo-react-gantt/editors/FormDateTimePicker.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as e from "react";
import { FieldWrapper as f } from "@progress/kendo-react-form";
import { Label as $, Hint as k, Error as w } from "@progress/kendo-react-labels";
import { DateTimePicker as y } from "@progress/kendo-react-dateinputs";
const _ = (p) => {
  const { validationMessage: a, touched: b, label: i, id: r, valid: o, disabled: l, hint: d, wrapperStyle: h, ...E } = p, t = b && a, s = !t && d, n = s ? `${r}_hint` : "", c = t ? `${r}_error` : "", m = i ? `${r}_label` : "";
  return /* @__PURE__ */ e.createElement(f, { style: h, className: "k-col-span-2" }, /* @__PURE__ */ e.createElement($, { id: m, editorId: r, editorValid: o, editorDisabled: l }, i), /* @__PURE__ */ e.createElement("div", { className: "k-form-field-wrap" }, /* @__PURE__ */ e.createElement(
    y,
    {
      ariaLabelledBy: m,
      ariaDescribedBy: `${n} ${c}`,
      valid: o,
      id: r,
      disabled: l,
      ...E
    }
  ), s && /* @__PURE__ */ e.createElement(k, { id: n }, d), t && /* @__PURE__ */ e.createElement(w, { id: c }, a)));
};
export {
  _ as FormDateTimePicker
};
