// @progress/kendo-react-gantt/messages/index.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
const t = "gantt.weekViewTitle", e = "gantt.dayViewTitle", a = "gantt.monthViewTitle", r = "gantt.yearViewTitle", n = "gantt.filterClearButton", o = "gantt.filterEqOperator", i = "gantt.filterNotEqOperator", s = "gantt.filterIsNullOperator", l = "gantt.filterIsNotNullOperator", d = "gantt.filterIsEmptyOperator", c = "gantt.filterIsNotEmptyOperator", p = "gantt.filterStartsWithOperator", g = "gantt.filterContainsOperator", f = "gantt.filterNotContainsOperator", T = "gantt.filterEndsWithOperator", u = "gantt.filterGteOperator", O = "gantt.filterGtOperator", S = "gantt.filterLteOperator", y = "gantt.filterLtOperator", E = "gantt.filterIsTrue", m = "gantt.filterIsFalse", I = "gantt.filterBooleanAll", k = "gantt.filterAfterOrEqualOperator", A = "gantt.filterAfterOperator", D = "gantt.filterBeforeOperator", b = "gantt.filterBeforeOrEqualOperator", h = "gantt.noRecords", C = "gantt.editSave", w = "gantt.editCancel", V = "gantt.editAdd", q = "gantt.editRemove", N = "gantt.editTabGeneral", F = "gantt.editTabPredecessors", v = "gantt.editTabSuccessors", P = "gantt.editorTitle", B = "gantt.editorTaskTitle", R = "gantt.editorTaskStart", L = "gantt.editorTaskEnd", G = "gantt.editorTaskPlannedStart", W = "gantt.editorTaskPlannedEnd", x = "gantt.editorTaskActualStart", M = "gantt.editorTaskActualEnd", Y = "gantt.editorTaskComplete", j = "gantt.editorTaskParent", z = "gantt.editorValidationRequired", H = "gantt.editorValidationStart", J = "gantt.editorValidationEnd", K = "gantt.editorValidationPercentCompleteRange", Q = "gantt.editorDependencyTypesFF", U = "gantt.editorDependencyTypesSF", X = "gantt.editorDependencyTypesFS", Z = "gantt.editorDependencyTypesSS", _ = "gantt.addTask", $ = "gantt.addChild", tt = "gantt.addAbove", et = "gantt.addBelow", at = "gantt.viewSelector", rt = "gantt.editorDelete", nt = "gantt.deleteConfirmation", ot = "gantt.deleteDialogTitle", it = "gantt.editorDependencyNameLabel", st = "gantt.editorDependencyTypeLabel", lt = {
  [t]: "Week",
  [e]: "Day",
  [a]: "Month",
  [r]: "Year",
  [n]: "Clear",
  [o]: "Is equal to",
  [i]: "Is not equal to",
  [s]: "Is null",
  [l]: "Is not null",
  [d]: "Is empty",
  [c]: "Is not empty",
  [p]: "Starts with",
  [g]: "Contains",
  [f]: "Does not contain",
  [T]: "Ends with",
  [u]: "Is greater than or equal to",
  [O]: "Is greater than",
  [S]: "Is less than or equal to",
  [y]: "Is less than",
  [E]: "Is true",
  [m]: "Is false",
  [I]: "(All)",
  [k]: "Is after or equal to",
  [A]: "Is after",
  [D]: "Is before",
  [b]: "Is before or equal to",
  [h]: "No records available",
  [C]: "Save",
  [w]: "Cancel",
  [V]: "Add",
  [q]: "Remove",
  [N]: "General",
  [F]: "Predecessors",
  [v]: "Successors",
  [P]: "Edit task",
  [B]: "Title",
  [R]: "Start",
  [L]: "End",
  [G]: "Planned Start Date",
  [W]: "Planned End Date",
  [x]: "Actual Start Date",
  [M]: "Actual End Date",
  [Y]: "Complete",
  [j]: "Parent",
  [z]: "Field is required.",
  [H]: "Start time must be be before End time.",
  [J]: "End time must be after Start time.",
  [K]: "Value should be between 0 and 100.",
  [Q]: "Finish-Finish",
  [X]: "Finish-Start",
  [U]: "Start-Finish",
  [Z]: "Start-Start",
  [_]: "Add Task",
  [$]: "Add Child",
  [tt]: "Add Above",
  [et]: "Add Below",
  [at]: "View Selector",
  [rt]: "Delete",
  [nt]: "Are you sure you want to delete this event?",
  [ot]: "Delete Event",
  [it]: "Name",
  [st]: "Type"
};
export {
  tt as addAbove,
  et as addBelow,
  $ as addChild,
  _ as addTask,
  e as dayViewTitle,
  nt as deleteConfirmation,
  ot as deleteDialogTitle,
  V as editorAdd,
  w as editorCancel,
  rt as editorDelete,
  it as editorDependencyNameLabel,
  st as editorDependencyTypeLabel,
  Q as editorDependencyTypesFF,
  X as editorDependencyTypesFS,
  U as editorDependencyTypesSF,
  Z as editorDependencyTypesSS,
  q as editorRemove,
  C as editorSave,
  N as editorTabGeneral,
  F as editorTabPredecessors,
  v as editorTabSuccessors,
  M as editorTaskActualEnd,
  x as editorTaskActualStart,
  Y as editorTaskComplete,
  L as editorTaskEnd,
  j as editorTaskParent,
  W as editorTaskPlannedEnd,
  G as editorTaskPlannedStart,
  R as editorTaskStart,
  B as editorTaskTitle,
  P as editorTitle,
  J as editorValidationEnd,
  K as editorValidationPercentCompleteRange,
  z as editorValidationRequired,
  H as editorValidationStart,
  A as filterAfterOperator,
  k as filterAfterOrEqualOperator,
  D as filterBeforeOperator,
  b as filterBeforeOrEqualOperator,
  I as filterBooleanAll,
  n as filterClearButton,
  g as filterContainsOperator,
  T as filterEndsWithOperator,
  o as filterEqOperator,
  O as filterGtOperator,
  u as filterGteOperator,
  d as filterIsEmptyOperator,
  m as filterIsFalse,
  c as filterIsNotEmptyOperator,
  l as filterIsNotNullOperator,
  s as filterIsNullOperator,
  E as filterIsTrue,
  y as filterLtOperator,
  S as filterLteOperator,
  f as filterNotContainsOperator,
  i as filterNotEqOperator,
  p as filterStartsWithOperator,
  lt as messages,
  a as monthViewTitle,
  h as noRecords,
  at as viewSelector,
  t as weekViewTitle,
  r as yearViewTitle
};
