// @progress/kendo-react-orgchart/client/ClientOrgChart.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as s from "react";
import o from "prop-types";
import { Navigation as k, classNames as $ } from "@progress/kendo-react-common";
import { loadTheme as z } from "../theming/theme-service.mjs";
import { ORGCHART_CLASSNAME as J } from "../utils/consts.mjs";
const Q = s.createContext({
  cardColors: void 0,
  onExpandChange: void 0,
  onKeyDown: void 0,
  onItemAction: void 0,
  onItemContextMenu: void 0,
  onItemDoubleClick: void 0,
  onGroupAction: void 0,
  onGroupBlur: void 0,
  onGroupFocus: void 0
}), A = (g) => {
  const n = {
    tabIndex: 0,
    navigatable: !0,
    expandField: "expanded",
    ariaLabel: "Org Chart",
    idField: "id",
    childrenField: "items",
    avatarField: "avatar",
    titleField: "title",
    subtitleField: "subtitle",
    ...g
  }, {
    navigatable: v,
    id: V,
    tabIndex: f,
    className: I,
    style: y,
    data: X,
    groupField: Y,
    cardWidth: Z,
    cardHeight: ee,
    groupTitleHeight: te,
    groupSubtitleHeight: ne,
    verticalLine: oe,
    idField: re,
    childrenField: ie,
    expandField: ae,
    hasChildrenField: ce,
    avatarField: le,
    titleField: se,
    subtitleField: de,
    onExpandChange: ue,
    onItemAction: h,
    onItemDoubleClick: ge,
    onItemContextMenu: he,
    onKeyDown: pe,
    onGroupAction: b,
    onGroupBlur: me,
    onGroupFocus: fe,
    cardsColors: x,
    ariaLabel: Fe,
    itemRender: Ce,
    groupTitleRender: ve,
    groupSubtitleRender: be,
    height: xe,
    children: Ee,
    ...D
  } = n, F = s.useRef(null), c = s.useRef(
    new k({
      tabIndex: f,
      root: F,
      selectors: [n.groupField ? ".k-orgchart-node-group-container" : ".k-orgchart-card"],
      focusOptions: {}
    })
  ), [w, G] = s.useState([]), d = s.useRef(null), N = (e, t, i) => {
    i.preventDefault(), t.focusNext(e);
  }, R = (e, t, i) => {
    i.preventDefault(), t.focusPrevious(e);
  }, S = (e, t, i, r) => {
    var l;
    if (i.preventDefault(), (n.groupField ? r.items.some((m) => m[n.expandField]) : r.item[n.expandField]) && n.onExpandChange)
      n.onExpandChange.call(void 0, r);
    else {
      const C = (l = e.closest(".k-orgchart-group")) == null ? void 0 : l.getAttribute("id"), p = document.querySelector(`[aria-owns="${C}"]`);
      p ? t.focusElement(p, e) : t.focusPrevious(e);
    }
  }, O = (e, t, i, r) => {
    var p;
    i.preventDefault();
    const a = n.groupField ? r.items.some((u) => u[n.expandField]) : r.item[n.expandField], l = n.groupField ? r.items.find((u) => u[n.childrenField]) : r.item[n.childrenField], m = n.groupField ? r.items.some((u) => n.hasChildrenField && u[n.hasChildrenField]) : n.hasChildrenField && r.item[n.hasChildrenField], C = n.groupField ? l && l[n.childrenField] && l[n.childrenField].length : l && l.length;
    if ((m || C) && !a && n.onExpandChange)
      n.onExpandChange.call(void 0, r);
    else {
      const u = n.groupField ? ".k-orgchart-node-group-container" : ".k-orgchart-card", W = e.getAttribute("aria-owns"), E = (p = document.getElementById(W)) == null ? void 0 : p.querySelector(u);
      E ? t.focusElement(E, e) : t.focusNext(e);
    }
  }, T = (e, t, i) => {
    i.preventDefault(), t.focusElement(t.first, null);
  }, H = (e, t, i) => {
    i.preventDefault(), t.focusElement(t.last, null);
  }, P = (e, t, i, r) => {
    i.stopPropagation();
    const a = e.className.indexOf("k-orgchart-node-group-container") !== -1;
    if (g.groupField && a) {
      d.current && d.current.first && d.current.focusElement(d.current.first, null);
      return;
    }
    n.onItemAction && n.onItemAction({
      event: i,
      item: r.item
    });
  }, L = (e) => {
    if (g.groupField) {
      const t = e.closest(".k-orgchart-node-group-container");
      t && c.current.focusElement(t, null);
    }
  };
  s.useEffect(() => {
    if (v && c.current.first && c.current.first.setAttribute("tabindex", String(f)), !x) {
      const e = F.current.ownerDocument || document;
      z((t) => {
        G(t);
      }, e);
    }
  }, [v, f]);
  const M = (e) => {
    if (n.navigatable) {
      const t = e.event, i = t.key === " " ? "Space" : t.key, r = g.groupField && t.target.className.indexOf("k-orgchart-card") !== -1 && d.current || c.current, a = t.target.closest(
        r.selectors.join(",")
      );
      switch (i) {
        case "ArrowUp":
          S(a, r, t, e);
          break;
        case "ArrowDown":
          O(a, r, t, e);
          break;
        case "ArrowLeft":
          R(a, r, t);
          break;
        case "ArrowRight":
          N(a, r, t);
          break;
        case "End":
          H(a, r, t);
          break;
        case "Home":
          T(a, r, t);
          break;
        case "Enter":
          P(a, r, t, e);
          break;
        case "Escape":
          L(a);
          break;
      }
    }
  }, B = (e) => {
    if (e.event.stopPropagation(), h && h.call(void 0, e), n.navigatable) {
      const t = e.event.target.closest(".k-orgchart-card");
      t && c.current.focusElement(t, null);
    }
  }, K = (e) => {
    e.event.stopPropagation(), h && h.call(void 0, e);
  }, _ = (e) => {
    e.event.stopPropagation(), h && h.call(void 0, e);
  }, j = (e) => {
    if (b && b.call(void 0, e), n.navigatable) {
      const t = e.event.target.closest(".k-orgchart-node-group-container");
      t && c.current.focusElement(t, null);
    }
  }, q = (e) => {
    d.current = null;
  }, U = (e) => {
    d.current = new k({
      tabIndex: 0,
      root: e.containerRef,
      selectors: [".k-orgchart-card"],
      focusOptions: {}
    });
  };
  return /* @__PURE__ */ s.createElement("div", { style: y, ref: F, className: $(I, J) }, /* @__PURE__ */ s.createElement(
    Q.Provider,
    {
      value: {
        cardColors: x || w,
        onExpandChange: g.onExpandChange,
        onKeyDown: M,
        onItemAction: B,
        onItemDoubleClick: _,
        onItemContextMenu: K,
        onGroupAction: j,
        onGroupBlur: q,
        onGroupFocus: U
      },
      ...D
    },
    g.children
  ));
};
A.propTypes = {
  className: o.string,
  style: o.object,
  id: o.string,
  ariaLabel: o.string,
  idField: o.string,
  childrenField: o.string,
  expandField: o.string,
  titleField: o.string,
  hasChildrenField: o.string,
  subtitleField: o.string,
  avatarField: o.string,
  cardsColors: o.array,
  data: o.array,
  groupField: o.string,
  cardHeight: o.number,
  cardWidth: o.number,
  groupTitleHeight: o.number,
  groupSubtitleHeight: o.number,
  verticalLine: o.number,
  height: o.oneOfType([o.string, o.number]),
  onExpandChange: o.func,
  onItemAction: o.func,
  onItemDoubleClick: o.func,
  onItemContextMenu: o.func,
  onGroupAction: o.func
};
A.displayName = "KendoClientOrgChart";
export {
  A as ClientOrgChart,
  Q as OrgChartEventsContextContext
};
