// @progress/kendo-react-orgchart/client/ExpandButton.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as o from "react";
import { Button as m } from "@progress/kendo-react-buttons";
import { minusIcon as p, plusIcon as i } from "@progress/kendo-svg-icons";
import { OrgChartEventsContextContext as x } from "./ClientOrgChart.mjs";
const C = (e) => {
  const { expanded: n, node: a, nodes: r } = e, t = o.useContext(x), s = (c) => {
    t.onExpandChange && t.onExpandChange.call(void 0, {
      event: c,
      expand: !n,
      item: a,
      items: r
    });
  };
  return /* @__PURE__ */ o.createElement(
    m,
    {
      className: "k-orgchart-button",
      icon: n ? "minus" : "plus",
      tabIndex: -1,
      svgIcon: n ? p : i,
      "aria-label": n ? "collapse" : "expand",
      onClick: s
    }
  );
};
export {
  C as ExpandButton
};
