// @progress/kendo-react-orgchart/client/OrgChartCard.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as n from "react";
import { Card as f } from "@progress/kendo-react-layout";
import { classNames as h } from "@progress/kendo-react-common";
import { OrgChartEventsContextContext as k } from "./ClientOrgChart.mjs";
const I = (r) => {
  const { node: o, children: i, style: a, ...s } = r, [l, c] = n.useState(!1), e = n.useContext(k), m = (t) => {
    e.onKeyDown && e.onKeyDown.call(void 0, {
      event: t,
      item: o
    });
  }, u = (t) => {
    e.onItemAction && e.onItemAction.call(void 0, {
      event: t,
      item: o
    });
  }, C = (t) => {
    e.onItemContextMenu && e.onItemContextMenu.call(void 0, {
      event: t,
      item: o
    });
  }, d = (t) => {
    e.onItemDoubleClick && e.onItemDoubleClick.call(void 0, {
      event: t,
      item: o
    });
  };
  return /* @__PURE__ */ n.createElement(
    f,
    {
      orientation: null,
      role: "treeitem",
      className: h("k-orgchart-card", { "k-focus": l }),
      onClick: u,
      onKeyDown: m,
      onContextMenu: C,
      onDoubleClick: d,
      style: a,
      "aria-selected": l,
      "aria-keyshortcuts": "Enter",
      onFocus: () => c(!0),
      onBlur: () => c(!1),
      ...s
    },
    i
  );
};
export {
  I as OrgChartCard
};
