// @progress/kendo-react-orgchart/client/OrgChartGroupContainer.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as n from "react";
import { classNames as h } from "@progress/kendo-react-common";
import { OrgChartEventsContextContext as C } from "./ClientOrgChart.mjs";
const x = (a) => {
  const { nodes: t, children: l, style: u, ...i } = a, [r, s] = n.useState(!1), o = n.useContext(C), c = n.useRef(null), m = (e) => {
    o.onKeyDown && o.onKeyDown({
      event: e,
      items: t
    });
  }, f = (e) => {
    o.onGroupAction && o.onGroupAction({
      event: e,
      items: t
    });
  }, p = (e) => {
    s(!0), o.onGroupFocus && o.onGroupFocus({
      event: e,
      items: t,
      containerRef: c
    });
  }, d = (e) => {
    s(!1), o.onGroupBlur && o.onGroupBlur({
      event: e,
      items: t
    });
  };
  return /* @__PURE__ */ n.createElement(
    "div",
    {
      role: "treeitem",
      "aria-selected": r,
      ref: c,
      className: h("k-orgchart-node-group-container", "k-vstack", { "k-focus": r }),
      style: u,
      onKeyDown: m,
      onClick: f,
      onFocus: p,
      onBlur: d,
      ...i
    },
    l
  );
};
export {
  x as OrgChartGroupContainer
};
