// @progress/kendo-react-scheduler/index.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import { Scheduler as r, schedulerDefaultProps as o } from "./Scheduler.mjs";
import { DATA_ACTION as u } from "./constants/index.mjs";
import { AgendaView as m, agendaViewDefaultProps as S } from "./views/agenda/AgendaView.mjs";
import { TimelineView as x, timeLineViewDefaultProps as h } from "./views/time/TimelineView.mjs";
import { DayView as p, dayViewDefaultProps as s } from "./views/day/DayView.mjs";
import { WeekView as a, weekViewDefaultProps as C } from "./views/week/WeekView.mjs";
import { WorkWeekView as I, workWeekDefaultProps as E } from "./views/week/WorkWeekView.mjs";
import { MonthView as w, monthViewDefaultProps as P } from "./views/month/MonthView.mjs";
import { SchedulerItem as O } from "./items/SchedulerItem.mjs";
import { SchedulerItemContent as _ } from "./items/SchedulerItemContent.mjs";
import { SchedulerViewItem as R, schedulerViewItemDefaultProps as g } from "./items/SchedulerViewItem.mjs";
import { SchedulerProportionalViewItem as N, schedulerProportionalViewItemDefaultProps as k } from "./items/SchedulerProportionalViewItem.mjs";
import { SchedulerEditItem as v, schedulerEditItemDefaultProps as H } from "./items/SchedulerEditItem.mjs";
import { SchedulerSlot as W, schedulerSlotDefaultProps as G } from "./slots/SchedulerSlot.mjs";
import { SchedulerViewSlot as U, schedulerViewSlotDefaultProps as y } from "./slots/SchedulerViewSlot.mjs";
import { SchedulerEditSlot as b, schedulerEditSlotDefaultProps as j } from "./slots/SchedulerEditSlot.mjs";
import { SchedulerTask as B } from "./tasks/SchedulerTask.mjs";
import { SchedulerViewTask as K, schedulerViewTaskDefaultProps as Q } from "./tasks/SchedulerViewTask.mjs";
import { SchedulerEditTask as Y, schedulerEditTaskDefaultProps as $ } from "./tasks/SchedulerEditTask.mjs";
import { SchedulerDrag as te, schedulerDragDefaultProps as re } from "./views/common/SchedulerDrag.mjs";
import { SchedulerResize as le, schedulerResizeDefaultProps as ue } from "./views/common/SchedulerResize.mjs";
import { SchedulerForm as me, schedulerFormDefaultProps as Se } from "./components/SchedulerForm.mjs";
import { SchedulerFormEditor as xe, schedulerFormEditorDefaultProps as he } from "./editors/SchedulerFormEditor.mjs";
import { SchedulerOccurrenceDialog as pe } from "./components/SchedulerOccurrenceDialog.mjs";
import { SchedulerRemoveDialog as fe } from "./components/SchedulerRemoveDialog.mjs";
import { SchedulerHeader as Ce } from "./components/header/SchedulerHeader.mjs";
import { DateHeaderCell as Ie, dateHeaderCellDefaultProps as Ee } from "./components/DateHeaderCell.mjs";
import { TimeHeaderCell as we, timeHeaderCellDefaultProps as Pe } from "./components/TimeHeaderCell.mjs";
import { SchedulerFooter as Oe } from "./components/footer/SchedulerFooter.mjs";
import { DRAG_ITEM_ACTION as _e } from "./items/hooks/use-drag-item.mjs";
import { ITEMS_SELECT_ACTION as Re } from "./hooks/use-items-selection.mjs";
import { REMOVE_ITEM_ACTION as Fe } from "./items/hooks/use-remove-item.mjs";
import { RESIZE_ITEM_ACTION as ke } from "./items/hooks/use-resize-item.mjs";
import { SERIES_ACTION as ve } from "./items/hooks/use-series.mjs";
import { SHOW_OCCURRENCE_DIALOG_ACTION as Le } from "./items/hooks/use-show-occurrence-dialog.mjs";
import { SHOW_REMOVE_DIALOG_ACTION as Ge } from "./items/hooks/use-show-remove-item-dialog.mjs";
import { SLOTS_SELECT_ACTION as Ue } from "./hooks/use-slots-selection.mjs";
import { SchedulerActiveViewContext as Ze, SchedulerDataContext as be, SchedulerDateContext as je, SchedulerDateRangeContext as qe, SchedulerFieldsContext as Be, SchedulerGroupsContext as Je, SchedulerOrientationContext as Ke, SchedulerPropsContext as Qe, SchedulerViewsContext as Xe, useSchedulerActiveViewContext as Ye, useSchedulerDataContext as $e, useSchedulerDateContext as et, useSchedulerDateRangeContext as tt, useSchedulerFieldsContext as rt, useSchedulerGroupsContext as ot, useSchedulerOrientationContext as lt, useSchedulerPropsContext as ut, useSchedulerViewsContext as dt } from "./context/SchedulerContext.mjs";
import { SchedulerEditItemDragItemContext as St, SchedulerEditItemFormItemContext as ct, SchedulerEditItemPropsContext as xt, SchedulerEditItemRemoveItemContext as ht, SchedulerEditItemResizeItemContext as it, SchedulerEditItemShowOccurrenceDialogContext as pt, SchedulerEditItemShowRemoveDialogContext as st, useSchedulerEditItemDragItemContext as ft, useSchedulerEditItemFormItemContext as at, useSchedulerEditItemPropsContext as Ct, useSchedulerEditItemRemoveItemContext as nt, useSchedulerEditItemResizeItemContext as It, useSchedulerEditItemShowOccurrenceDialogContext as Et, useSchedulerEditItemShowRemoveDialogContext as Dt } from "./context/SchedulerEditItemContext.mjs";
import { SchedulerEditSlotFormItemContext as Pt, SchedulerEditSlotPropsContext as Tt, useSchedulerEditSlotFormItemContext as Ot, useSchedulerEditSlotPropsContext as Vt } from "./context/SchedulerEditSlotContext.mjs";
import { ITEMS_FOCUS_ACTION as At } from "./hooks/use-items-focus.mjs";
import { SLOTS_FOCUS_ACTION as gt } from "./hooks/use-slots-focus.mjs";
export {
  m as AgendaView,
  u as DATA_ACTION,
  _e as DRAG_ITEM_ACTION,
  Ie as DateHeaderCell,
  p as DayView,
  At as ITEMS_FOCUS_ACTION,
  Re as ITEMS_SELECT_ACTION,
  w as MonthView,
  Fe as REMOVE_ITEM_ACTION,
  ke as RESIZE_ITEM_ACTION,
  ve as SERIES_ACTION,
  Le as SHOW_OCCURRENCE_DIALOG_ACTION,
  Ge as SHOW_REMOVE_DIALOG_ACTION,
  gt as SLOTS_FOCUS_ACTION,
  Ue as SLOTS_SELECT_ACTION,
  r as Scheduler,
  Ze as SchedulerActiveViewContext,
  be as SchedulerDataContext,
  je as SchedulerDateContext,
  qe as SchedulerDateRangeContext,
  te as SchedulerDrag,
  v as SchedulerEditItem,
  St as SchedulerEditItemDragItemContext,
  ct as SchedulerEditItemFormItemContext,
  xt as SchedulerEditItemPropsContext,
  ht as SchedulerEditItemRemoveItemContext,
  it as SchedulerEditItemResizeItemContext,
  pt as SchedulerEditItemShowOccurrenceDialogContext,
  st as SchedulerEditItemShowRemoveDialogContext,
  b as SchedulerEditSlot,
  Pt as SchedulerEditSlotFormItemContext,
  Tt as SchedulerEditSlotPropsContext,
  Y as SchedulerEditTask,
  Be as SchedulerFieldsContext,
  Oe as SchedulerFooter,
  me as SchedulerForm,
  xe as SchedulerFormEditor,
  Je as SchedulerGroupsContext,
  Ce as SchedulerHeader,
  O as SchedulerItem,
  _ as SchedulerItemContent,
  pe as SchedulerOccurrenceDialog,
  Ke as SchedulerOrientationContext,
  N as SchedulerProportionalViewItem,
  Qe as SchedulerPropsContext,
  fe as SchedulerRemoveDialog,
  le as SchedulerResize,
  W as SchedulerSlot,
  B as SchedulerTask,
  R as SchedulerViewItem,
  U as SchedulerViewSlot,
  K as SchedulerViewTask,
  Xe as SchedulerViewsContext,
  we as TimeHeaderCell,
  x as TimelineView,
  a as WeekView,
  I as WorkWeekView,
  S as agendaViewDefaultProps,
  Ee as dateHeaderCellDefaultProps,
  s as dayViewDefaultProps,
  P as monthViewDefaultProps,
  o as schedulerDefaultProps,
  re as schedulerDragDefaultProps,
  H as schedulerEditItemDefaultProps,
  j as schedulerEditSlotDefaultProps,
  $ as schedulerEditTaskDefaultProps,
  Se as schedulerFormDefaultProps,
  he as schedulerFormEditorDefaultProps,
  k as schedulerProportionalViewItemDefaultProps,
  ue as schedulerResizeDefaultProps,
  G as schedulerSlotDefaultProps,
  g as schedulerViewItemDefaultProps,
  y as schedulerViewSlotDefaultProps,
  Q as schedulerViewTaskDefaultProps,
  Pe as timeHeaderCellDefaultProps,
  h as timeLineViewDefaultProps,
  Ye as useSchedulerActiveViewContext,
  $e as useSchedulerDataContext,
  et as useSchedulerDateContext,
  tt as useSchedulerDateRangeContext,
  ft as useSchedulerEditItemDragItemContext,
  at as useSchedulerEditItemFormItemContext,
  Ct as useSchedulerEditItemPropsContext,
  nt as useSchedulerEditItemRemoveItemContext,
  It as useSchedulerEditItemResizeItemContext,
  Et as useSchedulerEditItemShowOccurrenceDialogContext,
  Dt as useSchedulerEditItemShowRemoveDialogContext,
  Ot as useSchedulerEditSlotFormItemContext,
  Vt as useSchedulerEditSlotPropsContext,
  rt as useSchedulerFieldsContext,
  ot as useSchedulerGroupsContext,
  lt as useSchedulerOrientationContext,
  ut as useSchedulerPropsContext,
  dt as useSchedulerViewsContext,
  C as weekViewDefaultProps,
  E as workWeekDefaultProps
};
