// @progress/kendo-react-pdf-viewer/messages.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
const e = "pdfviewer.zoomIn", o = "pdfviewer.zoomOut", t = "pdfviewer.zoomLevel", n = "pdfviewer.enableSelection", c = "pdfviewer.enablePanning", i = "pdfviewer.search", a = "pdfviewer.open", p = "pdfviewer.download", s = "pdfviewer.print", d = "pdfviewer.close", r = "pdfviewer.matchCase", f = "pdfviewer.prevMatch", l = "pdfviewer.nextMatch", v = "pdfviewer.actualWidth", w = "pdfviewer.fitToWidth", h = "pdfviewer.fitToPage", m = "pdfviewer.popupBlocked", u = "pdfviewer.searchOf", b = {
  [e]: "Zoom in",
  [o]: "Zoom out",
  [t]: "Choose zoom level",
  [n]: "Enable selection",
  [c]: "Enable panning",
  [i]: "Search",
  [a]: "Open",
  [p]: "Download",
  [u]: "of",
  [s]: "Print",
  [d]: "Close",
  [r]: "Match case",
  [f]: "Previous match",
  [l]: "Next match",
  [v]: "Actual width",
  [w]: "Fit to width",
  [h]: "Fit to page",
  [m]: "Popup is blocked."
};
export {
  v as actualWidth,
  d as close,
  p as download,
  c as enablePanning,
  n as enableSelection,
  h as fitToPage,
  w as fitToWidth,
  r as matchCase,
  b as messages,
  l as nextMatch,
  a as open,
  m as popupBlocked,
  f as prevMatch,
  s as print,
  i as search,
  u as searchOf,
  e as zoomIn,
  t as zoomLevel,
  o as zoomOut
};
