// @progress/kendo-react-conversational-ui/index.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import { ChatMessage as t } from "./components/ChatMessage.mjs";
import { Chat as m } from "./components/Chat.mjs";
import { SuggestionGroup as f } from "./chat/components/SuggestionsGroup.mjs";
import { HeroCard as a } from "./components/templates/HeroCard.mjs";
import { AIPrompt as u } from "./ai-prompt/AIPrompt.mjs";
import { PromptHeader as P } from "./ai-prompt/AIPromptHeader.mjs";
import { AIPromptFooter as s } from "./ai-prompt/AIPromptFooter.mjs";
import { AIPromptContent as w } from "./ai-prompt/AIPromptContent.mjs";
import { AIPromptView as d } from "./ai-prompt/views/AIPromptView.mjs";
import { AIPromptOutputView as l } from "./ai-prompt/views/AIPromptOutputView.mjs";
import { AIPromptCommandsView as D } from "./ai-prompt/views/AIPromptCommandsView.mjs";
import { AIPromptViewRender as H } from "./ai-prompt/views/AIPromptViewRender.mjs";
import { InlineAIPrompt as F } from "./inline-ai-prompt/InlineAIPrompt.mjs";
import { commandsViewDefaults as M, outputViewDefaults as O, promptViewDefaults as R } from "./ai-prompt/views/constants.mjs";
export {
  u as AIPrompt,
  D as AIPromptCommandsView,
  w as AIPromptContent,
  s as AIPromptFooter,
  l as AIPromptOutputView,
  d as AIPromptView,
  H as AIPromptViewRender,
  m as Chat,
  t as ChatMessage,
  a as HeroCard,
  F as InlineAIPrompt,
  P as PromptHeader,
  f as SuggestionGroup,
  M as commandsViewDefaults,
  O as outputViewDefaults,
  R as promptViewDefaults
};
