// @progress/kendo-react-taskboard/column/Column.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as t from "react";
import { classNames as a } from "@progress/kendo-react-common";
import { TASKBOARD_ITEM_TYPE as l, TASKBOARD_COLUMN as i, TASKBOARD_ITEM_ID as o } from "../constants.mjs";
import { TaskBoardTaskEditPane as n } from "../TaskBoardTaskEditPane.mjs";
const d = (e) => /* @__PURE__ */ t.createElement(t.Fragment, null, /* @__PURE__ */ t.createElement(
  "div",
  {
    ref: e.elementRef,
    style: e.style,
    className: a("k-taskboard-column", {
      "k-taskboard-column-edit": e.column.edit
    }),
    tabIndex: e.tabIndex,
    [l]: i,
    [o]: e.column.id
  },
  /* @__PURE__ */ t.createElement(
    e.header,
    {
      column: e.column,
      tasks: e.tasks,
      onTitleChange: e.onTitleChange,
      onColumnExitEdit: e.onColumnExitEdit,
      editButtonTitle: e.editButtonTitle,
      onColumnEnterEdit: e.onColumnEnterEdit,
      addButtonTitle: e.addButtonTitle,
      onShowAddCardDialog: e.onShowAddCardDialog,
      closeButtonTitle: e.closeButtonTitle,
      onColumnDelete: e.onColumnDelete
    }
  ),
  /* @__PURE__ */ t.createElement(e.body, null, e.children)
), e.showAddCard && /* @__PURE__ */ t.createElement(
  e.addCardDialog,
  {
    onClose: e.onCloseDialog,
    onSave: e.onTaskCreate,
    priorities: e.priorities,
    editPane: n
  }
), e.showEditCard && e.editedTask && /* @__PURE__ */ t.createElement(
  e.editCardDialog,
  {
    onClose: e.onCloseDialog,
    onSave: e.onTaskEdit,
    task: e.editedTask,
    priorities: e.priorities,
    editPane: n
  }
), e.showColumnConfirmDelete && /* @__PURE__ */ t.createElement(
  e.confirmDialog,
  {
    onClose: e.onColumnDelete,
    onConfirm: e.onColumnConfirmDelete,
    dialogMessage: e.confirmDialogMessage,
    dialogTitle: e.confirmDialogTitle,
    dialogConfirmButton: e.confirmDialogConfirmButton,
    dialogCancelButton: e.confirmDialogCancelButton
  }
));
d.displayName = "KendoReactTaskBoardColumn";
export {
  d as TaskBoardColumn
};
