// @progress/kendo-react-taskboard/column/ColumnHeader.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as t from "react";
import { Button as l } from "@progress/kendo-react-buttons";
import { Input as o } from "@progress/kendo-react-inputs";
import { classNames as c } from "@progress/kendo-react-common";
import { pencilIcon as i, plusIcon as m, xIcon as d } from "@progress/kendo-svg-icons";
const r = (e) => {
  const { edit: a, title: n } = e.column;
  return /* @__PURE__ */ t.createElement("div", { className: "k-taskboard-column-header" }, /* @__PURE__ */ t.createElement("div", { className: "k-taskboard-column-header-text k-text-ellipsis" }, a ? /* @__PURE__ */ t.createElement(
    o,
    {
      value: n,
      onChange: e.onTitleChange,
      onBlur: e.onColumnExitEdit,
      autoFocus: !0
    }
  ) : n), /* @__PURE__ */ t.createElement("span", { className: "k-spacer" }), /* @__PURE__ */ t.createElement("div", { className: c("k-taskboard-column-header-actions", { "k-disabled": a }) }, /* @__PURE__ */ t.createElement(
    l,
    {
      fillMode: "flat",
      icon: "pencil",
      svgIcon: i,
      title: e.editButtonTitle,
      onClick: e.onColumnEnterEdit
    }
  ), /* @__PURE__ */ t.createElement(
    l,
    {
      fillMode: "flat",
      icon: "plus",
      svgIcon: m,
      title: e.addButtonTitle,
      onClick: e.onShowAddCardDialog
    }
  ), /* @__PURE__ */ t.createElement(
    l,
    {
      fillMode: "flat",
      icon: "x",
      svgIcon: d,
      title: e.closeButtonTitle,
      onClick: e.onColumnDelete
    }
  )));
};
r.displayName = "KendoReactTaskBoardColumnHeader";
export {
  r as TaskBoardColumnHeader
};
