// @progress/kendo-react-taskboard/index.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
"use client";
import { TaskBoard as a } from "./TaskBoard.mjs";
import { TaskBoardToolbar as e } from "./TaskBoardToolbar.mjs";
import { TaskBoardCard as t } from "./card/Card.mjs";
import { TaskBoardCardHeader as T } from "./card/CardHeader.mjs";
import { TaskBoardCardBody as k } from "./card/CardBody.mjs";
import { TaskBoardColumn as p } from "./column/Column.mjs";
import { TaskBoardColumnHeader as C } from "./column/ColumnHeader.mjs";
import { TaskBoardColumnBody as l } from "./column/ColumnBody.mjs";
import { TaskBoardConfirmDialog as u } from "./TaskBoardConfirmDialog.mjs";
import { TaskBoardPreviewDialog as E } from "./card/PreviewDialog.mjs";
import { TaskBoardAddCard as D } from "./TaskBoardAddCard.mjs";
import { TaskBoardEditCard as P } from "./TaskBoardEditCard.mjs";
import { TaskBoardTaskEditPane as v } from "./TaskBoardTaskEditPane.mjs";
import { useTaskEditing as A } from "./hooks/taskEditing.mjs";
export {
  a as TaskBoard,
  D as TaskBoardAddCard,
  t as TaskBoardCard,
  k as TaskBoardCardBody,
  T as TaskBoardCardHeader,
  p as TaskBoardColumn,
  l as TaskBoardColumnBody,
  C as TaskBoardColumnHeader,
  u as TaskBoardConfirmDialog,
  P as TaskBoardEditCard,
  E as TaskBoardPreviewDialog,
  v as TaskBoardTaskEditPane,
  e as TaskBoardToolbar,
  A as useTaskEditing
};
