// @progress/kendo-react-taskboard/TaskBoardEditCard.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as T from "react";
import { useLocalization as E } from "@progress/kendo-react-intl";
import { taskBoardEditCardPaneTitle as i, messages as a, taskBoardAddCardTitleLabel as o, taskBoardAddCardDescriptionLabel as r, taskBoardAddCardPriorityLabel as n, taskBoardEditCardPaneSaveBtn as d, taskBoardEditPaneCancelBtn as g, taskBoardEditPaneTitleLabel as s, taskBoardEditPaneDescriptionLabel as l, taskBoardEditPanePriorityLabel as L } from "./messages/index.mjs";
import { useTaskEditing as P } from "./hooks/taskEditing.mjs";
const c = (e) => {
  const { onTitleChange: u, title: k, onDescriptionChange: B, description: p, onPriorityChange: C, priority: m, onSave: S } = P(e), t = E();
  return /* @__PURE__ */ T.createElement(
    e.editPane,
    {
      header: t.toLanguageString(i, a[i]) + e.task.title,
      titleInputTitle: t.toLanguageString(
        o,
        a[o]
      ),
      descriptionInputTitle: t.toLanguageString(
        r,
        a[r]
      ),
      priorityDropDownTitle: t.toLanguageString(
        n,
        a[n]
      ),
      task: e.task,
      saveButton: t.toLanguageString(
        d,
        a[d]
      ),
      cancelButton: t.toLanguageString(
        g,
        a[g]
      ),
      priorities: e.priorities,
      titleLabel: t.toLanguageString(
        s,
        a[s]
      ),
      descriptionLabel: t.toLanguageString(
        l,
        a[l]
      ),
      priorityLabel: t.toLanguageString(
        L,
        a[L]
      ),
      onSave: S,
      onClose: e.onClose,
      onTitleChange: u,
      title: k,
      onDescriptionChange: B,
      description: p,
      onPriorityChange: C,
      priority: m
    }
  );
};
c.propTypes = {};
c.displayName = "KendoReactTaskBoardEditCard";
export {
  c as TaskBoardEditCard
};
