// @progress/kendo-react-taskboard/TaskBoardAddCard.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as S from "react";
import { useLocalization as b } from "@progress/kendo-react-intl";
import { taskBoardAddCardPaneTitle as i, messages as a, taskBoardAddCardTitleLabel as o, taskBoardAddCardDescriptionLabel as r, taskBoardAddCardPriorityLabel as n, taskBoardEditPaneTitleLabel as d, taskBoardEditPaneDescriptionLabel as g, taskBoardEditPanePriorityLabel as s, taskBoardAddCardCreateButton as l, taskBoardEditPaneCancelBtn as L } from "./messages/index.mjs";
import { useTaskEditing as P } from "./hooks/taskEditing.mjs";
const c = (e) => {
  const { onTitleChange: p, title: u, onDescriptionChange: B, description: C, onPriorityChange: k, priority: m, onSave: T } = P(e), t = b();
  return /* @__PURE__ */ S.createElement(
    e.editPane,
    {
      header: t.toLanguageString(i, a[i]),
      titleInputTitle: t.toLanguageString(
        o,
        a[o]
      ),
      descriptionInputTitle: t.toLanguageString(
        r,
        a[r]
      ),
      priorityDropDownTitle: t.toLanguageString(
        n,
        a[n]
      ),
      titleLabel: t.toLanguageString(
        d,
        a[d]
      ),
      descriptionLabel: t.toLanguageString(
        g,
        a[g]
      ),
      priorityLabel: t.toLanguageString(
        s,
        a[s]
      ),
      saveButton: t.toLanguageString(
        l,
        a[l]
      ),
      cancelButton: t.toLanguageString(
        L,
        a[L]
      ),
      priorities: e.priorities,
      onSave: T,
      onClose: e.onClose,
      onTitleChange: p,
      title: u,
      onDescriptionChange: B,
      description: C,
      onPriorityChange: k,
      priority: m
    }
  );
};
c.propTypes = {};
c.displayName = "KendoReactTaskBoardAddCard";
export {
  c as TaskBoardAddCard
};
