// @progress/kendo-react-taskboard/TaskBoardConfirmDialog.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as e from "react";
import o from "prop-types";
import { Dialog as g, DialogActionsBar as c } from "@progress/kendo-react-dialogs";
import { Button as i } from "@progress/kendo-react-buttons";
const r = (t) => {
  const { onClose: n, onConfirm: a, dialogMessage: l, dialogTitle: s, dialogConfirmButton: m, dialogCancelButton: d } = t;
  return /* @__PURE__ */ e.createElement(g, { title: s, closeIcon: !1 }, l, /* @__PURE__ */ e.createElement(c, { layout: "end" }, /* @__PURE__ */ e.createElement(i, { themeColor: "primary", onClick: a }, m), /* @__PURE__ */ e.createElement(i, { onClick: n }, d)));
};
r.propTypes = {
  onClose: o.func.isRequired,
  onConfirm: o.func.isRequired,
  dialogMessage: o.string.isRequired,
  dialogTitle: o.string.isRequired,
  dialogConfirmButton: o.string.isRequired,
  dialogCancelButton: o.string.isRequired
};
r.displayName = "KendoReactTaskBoardConfirmDialog";
export {
  r as TaskBoardConfirmDialog
};
