// @progress/kendo-react-taskboard/messages/index.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
const t = "taskBoard.toolbarAddColumnButton", a = "taskBoard.toolbarSearchField", e = "taskBoard.editColumnButton", o = "taskBoard.addCardButton", d = "taskBoard.deleteColumnButton", r = "taskBoard.editCardButton", n = "taskBoard.deleteCardButton", s = "taskBoard.deleteTaskDialogMessage", l = "taskBoard.deleteTaskDialogTitle", i = "taskBoard.deleteTaskConfirmButton", B = "taskBoard.deleteTaskCancelButton", k = "taskBoard.addCardTitle", c = "taskBoard.addCardTitleLabel", C = "taskBoard.addCardDescriptionLabel", u = "taskBoard.addCardPriorityLabel", D = "taskBoard.addCardCreateButton", P = "taskBoard.deleteColumnDialogMessage", m = "taskBoard.deleteColumnDialogTitle", T = "taskBoard.deleteColumnConfirmButton", b = "taskBoard.deleteColumnCancelButton", g = "taskBoard.editCardPaneTitle", E = "taskBoard.editCardPaneSaveButton", L = "taskBoard.editPaneTitleLabel", y = "taskBoard.editPaneDescriptionLabel", A = "taskBoard.editPanePriorityLabel", p = "taskBoard.editPaneCancelButton", v = "taskBoard.previewPanePriorityLabel", w = "taskBoard.previewPaneDeleteButton", h = "taskBoard.previewPaneEditButton", S = {
  [t]: "Add column",
  [a]: "Search",
  [e]: "Edit column",
  [o]: "Add card",
  [d]: "Delete column",
  [r]: "Edit card",
  [n]: "Delete card",
  [s]: "Are you sure you want to delete this card?",
  [l]: "Delete Card",
  [i]: "Delete",
  [B]: "Cancel",
  [P]: "Are you sure you want to delete this column?",
  [m]: "Delete column?",
  [T]: "Delete",
  [b]: "Cancel",
  [k]: "Create new card",
  [D]: "Create",
  [p]: "Cancel",
  [g]: "Edit",
  [E]: "Save changes",
  [L]: "Title:",
  [y]: "Description:",
  [A]: "Priority:",
  [c]: "Title",
  [C]: "Description",
  [u]: "Priority",
  [v]: "Priority:",
  [w]: "Delete",
  [h]: "Edit"
};
export {
  S as messages,
  o as taskBoardAddCardButton,
  D as taskBoardAddCardCreateButton,
  C as taskBoardAddCardDescriptionLabel,
  k as taskBoardAddCardPaneTitle,
  u as taskBoardAddCardPriorityLabel,
  c as taskBoardAddCardTitleLabel,
  t as taskBoardAddColumnButton,
  b as taskBoardDelColumnCancelBtn,
  T as taskBoardDelColumnConfirmBtn,
  P as taskBoardDelColumnDialogMsg,
  m as taskBoardDelColumnDialogTitle,
  n as taskBoardDeleteCardButton,
  d as taskBoardDeleteColumnButton,
  B as taskBoardDeleteTaskCancelButton,
  i as taskBoardDeleteTaskConfirmButton,
  s as taskBoardDeleteTaskDialogMessage,
  l as taskBoardDeleteTaskDialogTitle,
  r as taskBoardEditCardButton,
  E as taskBoardEditCardPaneSaveBtn,
  g as taskBoardEditCardPaneTitle,
  e as taskBoardEditColumnButton,
  p as taskBoardEditPaneCancelBtn,
  y as taskBoardEditPaneDescriptionLabel,
  A as taskBoardEditPanePriorityLabel,
  L as taskBoardEditPaneTitleLabel,
  w as taskBoardPreviewPaneDeleteBtn,
  h as taskBoardPreviewPaneEditBtn,
  v as taskBoardPreviewPanePriorityLabel,
  a as taskBoardSearchField
};
