// @progress/kendo-react-pivotgrid/components/Row.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as t from "react";
import { useMouse as i, classNames as s } from "@progress/kendo-react-common";
const o = t.forwardRef((e, n) => {
  const a = t.useRef(null), r = t.useRef(null);
  t.useImperativeHandle(a, () => ({
    element: r.current,
    props: e
  })), t.useImperativeHandle(n, () => a.current);
  const l = i(e, a);
  return /* @__PURE__ */ t.createElement(
    "tr",
    {
      ref: r,
      ...l,
      id: e.id,
      style: e.style,
      tabIndex: e.tabIndex,
      children: e.children,
      role: e.role,
      "aria-owns": e.ariaOwns,
      className: s("k-pivotgrid-row", e.className)
    }
  );
});
o.displayName = "KendoReactPivotGridRow";
export {
  o as PivotGridRow
};
