// @progress/kendo-react-gantt/context/GanttContext.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as e from "react";
import { ZonedDate as m } from "@progress/kendo-date-math";
import { noop as t } from "@progress/kendo-react-common";
import { DEFAULT_TASK_MODEL_FIELDS as N, DEFAULT_DEPENDENCY_MODEL_FIELDS as h } from "../constants/index.mjs";
const n = e.createContext([]), H = () => e.useContext(n);
n.displayName = "KendoReactGanttTaskDataContext";
const o = e.createContext([]), f = () => e.useContext(o);
o.displayName = "KendoReactGanttDependencyDataContext";
const a = e.createContext(N), A = () => e.useContext(a);
a.displayName = "KendoReactGanttTaskModelFieldsContext";
const s = e.createContext(
  h
), V = () => e.useContext(s);
s.displayName = "KendoReactGanttDependencyModelFieldsContext";
const c = e.createContext({}), _ = () => e.useContext(c);
c.displayName = "KendoReactGanttPropsContext";
const C = e.createContext({
  start: /* @__PURE__ */ new Date(),
  end: /* @__PURE__ */ new Date(),
  zonedStart: m.fromLocalDate(/* @__PURE__ */ new Date()),
  zonedEnd: m.fromLocalDate(/* @__PURE__ */ new Date())
}), b = () => e.useContext(C);
C.displayName = "KendoReactGanttDateRangeContext";
const r = e.createContext([]), z = () => e.useContext(r);
r.displayName = "KendoReactGanttViewsContext";
const l = e.createContext(["week", t]), I = () => e.useContext(l);
l.displayName = "KendoReactGanttActiveViewContext";
const x = e.createContext(0), O = () => e.useContext(x);
x.displayName = "KendoReactGanttRowHeightContext";
const d = e.createContext(0), U = () => e.useContext(d);
d.displayName = "KendoReactGanttToolbarHeightContext";
const i = e.createContext({
  onDataStateChange: t,
  onSortChange: t,
  onFilterChange: t,
  onColumnMenuFilterChange: t,
  onExpandChange: t,
  onRowClick: t,
  onRowDoubleClick: t,
  onRowContextMenu: t,
  onColumnResize: t,
  onColumnReorder: t,
  onDependencyCreate: t,
  onKeyDown: t,
  onSelectionChange: t,
  onHeaderSelectionChange: t
}), Y = () => e.useContext(i);
i.displayName = "KendoReactGanttEventsContext";
const u = e.createContext({
  onTaskClick: t,
  onTaskDoubleClick: t,
  onTaskContextMenu: t
}), Z = () => e.useContext(u);
u.displayName = "KendoReactGanttTaskEventsContext";
const F = ({
  children: G,
  taskData: D,
  dependencyData: v,
  taskModelFields: p,
  dependencyModelFields: E,
  props: R,
  dateRange: y,
  views: w,
  activeView: T,
  rowHeight: k,
  toolbarHeight: g,
  events: P,
  taskEvents: K
}) => /* @__PURE__ */ e.createElement(c.Provider, { value: R }, /* @__PURE__ */ e.createElement(n.Provider, { value: D }, /* @__PURE__ */ e.createElement(o.Provider, { value: v }, /* @__PURE__ */ e.createElement(a.Provider, { value: p }, /* @__PURE__ */ e.createElement(s.Provider, { value: E }, /* @__PURE__ */ e.createElement(C.Provider, { value: y }, /* @__PURE__ */ e.createElement(r.Provider, { value: w }, /* @__PURE__ */ e.createElement(l.Provider, { value: T }, /* @__PURE__ */ e.createElement(x.Provider, { value: k }, /* @__PURE__ */ e.createElement(d.Provider, { value: g }, /* @__PURE__ */ e.createElement(i.Provider, { value: P }, /* @__PURE__ */ e.createElement(u.Provider, { value: K }, G))))))))))));
F.displayName = "KendoReactGanttContext";
export {
  l as GanttActiveViewContext,
  F as GanttContext,
  C as GanttDateRangeContext,
  o as GanttDependencyDataContext,
  s as GanttDependencyModelFieldsContext,
  i as GanttEventsContext,
  c as GanttPropsContext,
  x as GanttRowHeightContext,
  n as GanttTaskDataContext,
  u as GanttTaskEventsContext,
  a as GanttTaskModelFieldsContext,
  d as GanttToolbarHeightContext,
  r as GanttViewsContext,
  I as useGanttActiveViewContext,
  b as useGanttDateRangeContext,
  f as useGanttDependencyDataContext,
  V as useGanttDependencyModelFieldsContext,
  Y as useGanttEventsContext,
  _ as useGanttPropsContext,
  O as useGanttRowHeightContext,
  H as useGanttTaskDataContext,
  Z as useGanttTaskEventsContext,
  A as useGanttTaskModelFieldsContext,
  U as useGanttToolbarHeightContext,
  z as useGanttViewsContext
};
