// @progress/kendo-react-gantt/context/GanttViewContext.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as e from "react";
import { noop as t } from "@progress/kendo-react-common";
const n = e.createContext([{ current: {} }, t]), x = () => e.useContext(n);
n.displayName = "KendoReactGanttViewContext";
const a = e.createContext(0), C = () => e.useContext(a);
a.displayName = "KendoReactGanttViewTimelineWidthContext";
const o = e.createContext({
  onDependencyHandlePress: t,
  onDependencyHandleDrag: t,
  onDependencyHandleRelease: t,
  isEnabled: !1,
  draggedId: null
}), m = () => e.useContext(o);
o.displayName = "KendoReactGanttViewDependencyDragContext";
const c = ({
  children: i,
  tasksStore: s,
  timelineWidth: r,
  dependencyDrag: d
}) => /* @__PURE__ */ e.createElement(n.Provider, { value: s }, /* @__PURE__ */ e.createElement(a.Provider, { value: r }, /* @__PURE__ */ e.createElement(o.Provider, { value: d }, i)));
c.displayName = "KendoReactGanttViewContext";
export {
  c as GanttViewContext,
  o as GanttViewDependencyDragContext,
  n as GanttViewTasksContext,
  a as GanttViewTimelineWidthContext,
  m as useGanttViewDependencyDragContext,
  x as useGanttViewTasksContext,
  C as useGanttViewTimelineWidthContext
};
