// @progress/kendo-react-editor/messages/index.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
const e = "editor.bold", t = "editor.italic", o = "editor.underline", r = "editor.strikethrough", i = "editor.subscript", l = "editor.superscript", n = "editor.hyperlink", s = "editor.insertFile", d = "editor.print", c = "editor.selectAll", a = "editor.cleanFormatting", p = "editor.pdf", b = "editor.foregroundColor", P = "editor.backgroundColor", g = "editor.createTable", C = "editor.insertTableHint", h = "editor.addRowBefore", m = "editor.addRowAfter", f = "editor.addColumnBefore", R = "editor.addColumnAfter", A = "editor.deleteRow", u = "editor.deleteColumn", y = "editor.deleteTable", T = "editor.mergeCells", S = "editor.splitCell", B = "editor.hyperlink-dialog-title", I = "editor.hyperlink-dialog-content-address", w = "editor.hyperlink-dialog-content-title", H = "editor.hyperlink-dialog-content-newwindow", k = "editor.hyperlink-dialog-cancel", v = "editor.hyperlink-dialog-insert", W = "editor.insertfile-dialog-title", L = "editor.insertfile-dialog-content-address", F = "editor.insertfile-dialog-content-title", x = "editor.insertfile-dialog-cancel", M = "editor.insertfile-dialog-insert", D = "editor.image", N = "editor.image-dialog-title", U = "editor.image-address", E = "editor.image-title", G = "editor.image-altText", z = "editor.image-width", O = "editor.image-height", J = "editor.image-cancel", V = "editor.image-insert", j = "editor.viewHtml", q = "editor.viewHtml-dialog-title", K = "editor.viewHtml-cancel", Q = "editor.viewHtml-update", X = "editor.unlink", Y = "editor.undo", Z = "editor.redo", _ = "editor.fontSize", $ = "editor.fontName", ee = "editor.format", te = "editor.alignLeft", oe = "editor.alignRight", re = "editor.alignCenter", ie = "editor.alignJustify", le = "editor.indent", ne = "editor.outdent", se = "editor.orderedList", de = "editor.bulletList", ce = "editor.findReplace-tool-title", ae = "editor.findReplace-dialog-title", pe = "editor.findReplace-tab-find", be = "editor.findReplace-tab-replace", Pe = "editor.findReplace-findWhat", ge = "editor.findReplace-replaceWith", Ce = "editor.findReplace-replace", he = "editor.findReplace-replaceAll", me = "editor.findReplace-matchCase", fe = "editor.findReplace-matchWord", Re = "editor.findReplace-matchCyclic", Ae = "editor.findReplace-useRegExp", ue = "editor.findReplace-prevMatch", ye = "editor.findReplace-nextMatch", Te = "editor.findReplace-matches", Se = "editor.findReplace-invalidRegExp", Be = "editor.iframeTitle", Ie = "editorTools.foreColorLabel", we = "editorTools.backColorLabel", He = "editor.cellProperties", ke = "editor.cellPropertiesLeftTop", ve = "editor.cellPropertiesCenterTop", We = "editor.cellPropertiesRightTop", Le = "editor.cellPropertiesLeftMiddle", Fe = "editor.cellPropertiesCenterMiddle", xe = "editor.cellPropertiesRightMiddle", Me = "editor.cellPropertiesLeftBottom", De = "editor.cellPropertiesCenterBottom", Ne = "editor.cellPropertiesRightBottom", Ue = "editor.cellPropertiesNoAlignment", Ee = "editor.cellPropertiesWrap", Ge = "editor.cellPropertiesFitToCell", ze = "editor.cellPropertiesApplyToAllCells", Oe = "editor.cellPropertiesWidth", Je = "editor.cellPropertiesApplyToColumn", Ve = "editor.cellPropertiesHeight", je = "editor.cellPropertiesApplyToRow", qe = "editor.cellPropertiesAlignment", Ke = "editor.cellPropertiesTextControl", Qe = "editor.cellPropertiesCellPadding", Xe = "editor.cellPropertiesBackground", Ye = "editor.cellPropertiesBorderWidth", Ze = "editor.cellPropertiesBorderColor", _e = "editor.cellPropertiesBorderStyle", $e = "editor.cellPropertiesId", et = "editor.cellPropertiesCssClass", tt = "editor.cellPropertiesSave", ot = "editor.cellPropertiesCancel", rt = "editor.tableProperties", it = "editor.tablePropertiesAlignLeft", lt = "editor.tablePropertiesAlignCenter", nt = "editor.tablePropertiesAlignRight", st = "editor.tablePropertiesNoAlignment", dt = "editor.tablePropertiesLeft", ct = "editor.tablePropertiesCenter", at = "editor.tablePropertiesRight", pt = "editor.tablePropertiesTop", bt = "editor.tablePropertiesBottom", Pt = "editor.tablePropertiesNone", gt = "editor.tablePropertiesUsingScopeAttribute", Ct = "editor.tablePropertiesUsingIdAttributes", ht = "editor.tablePropertiesRows", mt = "editor.tablePropertiesColumns", ft = "editor.tablePropertiesWidth", Rt = "editor.tablePropertiesHeight", At = "editor.tablePropertiesPosition", ut = "editor.tablePropertiesAlignment", yt = "editor.tablePropertiesCellSpacing", Tt = "editor.tablePropertiesCellPadding", St = "editor.tablePropertiesTableBackground", Bt = "editor.tablePropertiesBorderWidth", It = "editor.tablePropertiesBorderColor", wt = "editor.tablePropertiesBorderStyle", Ht = "editor.tablePropertiesCollapseBorders", kt = "editor.tablePropertiesID", vt = "editor.tablePropertiesCssClass", Wt = "editor.tablePropertiesAccessibility", Lt = "editor.tablePropertiesCaption", Ft = "editor.tablePropertiesCaptionPosition", xt = "editor.tablePropertiesCaptionAlignment", Mt = "editor.tablePropertiesHeaderRows", Dt = "editor.tablePropertiesHeaderCols", Nt = "editor.tablePropertiesAssociateHeaders", Ut = "editor.tablePropertiesGeneral", Et = "editor.tablePropertiesAdvanced", Gt = "editor.tablePropertiesSave", zt = "editor.tablePropertiesCancel", Ot = "editor.borderStyleNone", Jt = "editor.borderStyleSolid", Vt = "editor.borderStyleDotted", jt = "editor.borderStyleDashed", qt = "editor.borderStyleDouble", Kt = "editor.borderStyleGroove", Qt = "editor.borderStyleRidge", Xt = "editor.borderStyleInset", Yt = "editor.borderStyleOutset", Zt = "editor.borderStyleInitial", _t = "editor.borderStyleInherit", $t = "editor.borderStyleHidden", eo = "editor.sizeAuto", to = {
  [e]: "Bold",
  [t]: "Italic",
  [o]: "Underline",
  [r]: "Strikethrough",
  [i]: "Subscript",
  [l]: "Superscript",
  [X]: "Remove hyperlink",
  [Y]: "Undo",
  [Z]: "Redo",
  [_]: "Font Size",
  [$]: "Font Name",
  [ee]: "Format",
  [te]: "Align text left",
  [oe]: "Align text right",
  [re]: "Center text",
  [ie]: "Justify",
  [le]: "Indent",
  [ne]: "Outdent",
  [se]: "Insert ordered list",
  [de]: "Insert unordered list",
  [d]: "Print",
  [c]: "Select All",
  [a]: "Clean formatting",
  [p]: "Export as PDF",
  [b]: "Color",
  [P]: "Background color",
  // Tables
  [g]: "Create a table",
  [C]: "Create a {0} x {1} table",
  [h]: "Add row above",
  [m]: "Add row below",
  [f]: "Add column on the left",
  [R]: "Add column on the right",
  [A]: "Delete row",
  [u]: "Delete column",
  [y]: "Delete table",
  [T]: "Merge cells",
  [S]: "Split cell",
  // Insert Link
  [n]: "Insert hyperlink",
  [B]: "Insert hyperlink",
  [I]: "Web address",
  [w]: "Title",
  [H]: "Open link in new window",
  [k]: "Cancel",
  [v]: "Insert",
  // Insert File
  [s]: "Insert file",
  [W]: "Insert file",
  [L]: "Web address",
  [F]: "Title",
  [x]: "Cancel",
  [M]: "Insert",
  // Insert Image
  [D]: "Insert image",
  [N]: "Insert image",
  [U]: "Web address",
  [E]: "Title",
  [G]: "Alternate text",
  [z]: "Width (px)",
  [O]: "Height (px)",
  [J]: "Cancel",
  [V]: "Insert",
  // View HTML
  [j]: "View HTML",
  [q]: "View HTML",
  [K]: "Cancel",
  [Q]: "Update",
  // Find and Replace
  [ce]: "Find and Replace",
  [ae]: "Find and Replace",
  [pe]: "Find",
  [be]: "Replace",
  [Pe]: "Find What:",
  [ge]: "Replace With",
  [Ce]: "Replace",
  [he]: "Replace All",
  [me]: "Match Case",
  [fe]: "Match whole word only",
  [Re]: "Match cyclic (Wrap around)",
  [Ae]: "Regular Expression",
  [ue]: "Prev",
  [ye]: "Next",
  [Te]: "{0} of {1} matches",
  [Se]: "Invalid regular expression.",
  [Be]: "Editable area. Press Alt + F10 for toolbar.",
  [Ie]: "Fore Color",
  [we]: "Back Color",
  [He]: "Cell properties",
  [ke]: "Left Top",
  [ve]: "Center Top",
  [We]: "Right Top",
  [Le]: "Left Middle",
  [Fe]: "Center Middle",
  [xe]: "Right Middle",
  [Me]: "Left Bottom",
  [De]: "Center Bottom",
  [Ne]: "Right Bottom",
  [Ue]: "No Alignment",
  [Ee]: "Wrap",
  [Ge]: "Fit to cell",
  [ze]: "Apply to all cells",
  [Oe]: "Width",
  [Je]: "apply to column",
  [Ve]: "Height",
  [je]: "apply to row",
  [qe]: "Alignment",
  [Ke]: "Text control",
  [Qe]: "Cell padding",
  [Xe]: "Background",
  [Ye]: "Border width",
  [Ze]: "Border color",
  [_e]: "Border style",
  [$e]: "ID",
  [et]: "CSS class",
  [tt]: "Save",
  [ot]: "Cancel",
  [rt]: "Table properties",
  [it]: "Align Left",
  [lt]: "Align Center",
  [nt]: "Align Right",
  [st]: "No Alignment",
  [dt]: "Left",
  [ct]: "Center",
  [at]: "Right",
  [pt]: "Top",
  [bt]: "Bottom",
  [Pt]: "None",
  // eslint-disable-next-line @typescript-eslint/quotes
  [gt]: "Using 'scope' attribute",
  // eslint-disable-next-line @typescript-eslint/quotes
  [Ct]: "Using 'id' attributes",
  [ht]: "Rows",
  [mt]: "Columns",
  [ft]: "Width",
  [Rt]: "Height",
  [At]: "Position",
  [ut]: "Alignment",
  [yt]: "Cell spacing",
  [Tt]: "Cell padding",
  [St]: "Table background",
  [Bt]: "Border width",
  [It]: "Border color",
  [wt]: "Border style",
  [Ht]: "Collapse borders",
  [kt]: "ID",
  [vt]: "CSS class",
  [Wt]: "Accessibility",
  [Lt]: "Caption",
  [Ft]: "Position",
  [xt]: "Alignment",
  [Mt]: "Header rows",
  [Dt]: "Header cols",
  [Nt]: "Associate headers",
  [Ut]: "General",
  [Et]: "Advanced",
  [Gt]: "Save",
  [zt]: "Cancel",
  [Ot]: "None",
  [Jt]: "Solid",
  [Vt]: "Dotted",
  [jt]: "Dashed",
  [qt]: "Double",
  [Kt]: "Groove",
  [Qt]: "Ridge",
  [Xt]: "Inset",
  [Yt]: "Outset",
  [Zt]: "Initial",
  [_t]: "Inherit",
  [$t]: "Hidden",
  [eo]: "Auto"
}, oo = {
  bold: e,
  italic: t,
  underline: o,
  strikethrough: r,
  subscript: i,
  superscript: l,
  unlink: X,
  undo: Y,
  redo: Z,
  fontSize: _,
  fontName: $,
  format: ee,
  alignLeft: te,
  alignRight: oe,
  alignCenter: re,
  alignJustify: ie,
  indent: le,
  outdent: ne,
  orderedList: se,
  bulletList: de,
  print: d,
  selectAll: c,
  cleanFormatting: a,
  pdf: p,
  foregroundColor: b,
  backgroundColor: P,
  createTable: g,
  createTableHint: C,
  addRowBefore: h,
  addRowAfter: m,
  addColumnBefore: f,
  addColumnAfter: R,
  deleteRow: A,
  deleteColumn: u,
  deleteTable: y,
  mergeCells: T,
  splitCell: S,
  insertHyperlink: n,
  insertHyperlinkDialogTitle: B,
  insertHyperlinkAddress: I,
  insertHyperlinkTitle: w,
  insertHyperlinkNewWindow: H,
  insertHyperlinkCancel: k,
  insertHyperlinkInsert: v,
  insertFile: s,
  insertFileDialogTitle: W,
  insertFileAddress: L,
  insertFileTitle: F,
  insertFileCancel: x,
  insertFileInsert: M,
  insertImage: D,
  insertImageDialogTitle: N,
  insertImageAddress: U,
  insertImageTitle: E,
  insertImageAltText: G,
  insertImageWidth: z,
  insertImageHeight: O,
  insertImageCancel: J,
  insertImageInsert: V,
  viewHtml: j,
  viewHtmlDialogTitle: q,
  viewHtmlCancel: K,
  viewHtmlUpdate: Q,
  findReplaceToolTitle: ce,
  findReplaceDialogTitle: ae,
  findReplaceTabFind: pe,
  findReplaceTabReplace: be,
  findReplaceFindWhat: Pe,
  findReplaceReplaceWith: ge,
  findReplaceReplace: Ce,
  findReplaceReplaceAll: he,
  findReplaceMatchCase: me,
  findReplaceMatchWord: fe,
  findReplaceMatchCyclic: Re,
  findReplaceUseRegExp: Ae,
  findReplacePrevMatch: ue,
  findReplaceNextMatch: ye,
  findReplaceMatches: Te,
  findReplaceInvalidRegExp: Se,
  iframeTitle: Be,
  foreColorLabel: Ie,
  backColorLabel: we,
  cellProperties: He,
  cellPropertiesLeftTop: ke,
  cellPropertiesCenterTop: ve,
  cellPropertiesRightTop: We,
  cellPropertiesLeftMiddle: Le,
  cellPropertiesCenterMiddle: Fe,
  cellPropertiesRightMiddle: xe,
  cellPropertiesLeftBottom: Me,
  cellPropertiesCenterBottom: De,
  cellPropertiesRightBottom: Ne,
  cellPropertiesNoAlignment: Ue,
  cellPropertiesWrap: Ee,
  cellPropertiesFitToCell: Ge,
  cellPropertiesApplyToAllCells: ze,
  cellPropertiesWidth: Oe,
  cellPropertiesApplyToColumn: Je,
  cellPropertiesHeight: Ve,
  cellPropertiesApplyToRow: je,
  cellPropertiesAlignment: qe,
  cellPropertiesTextControl: Ke,
  cellPropertiesCellPadding: Qe,
  cellPropertiesBackground: Xe,
  cellPropertiesBorderWidth: Ye,
  cellPropertiesBorderColor: Ze,
  cellPropertiesBorderStyle: _e,
  cellPropertiesId: $e,
  cellPropertiesCssClass: et,
  cellPropertiesSave: tt,
  cellPropertiesCancel: ot,
  tableProperties: rt,
  tablePropertiesAlignLeft: it,
  tablePropertiesAlignCenter: lt,
  tablePropertiesAlignRight: nt,
  tablePropertiesNoAlignment: st,
  tablePropertiesLeft: dt,
  tablePropertiesCenter: ct,
  tablePropertiesRight: at,
  tablePropertiesTop: pt,
  tablePropertiesBottom: bt,
  tablePropertiesNone: Pt,
  tablePropertiesUsingScopeAttribute: gt,
  tablePropertiesUsingIdAttributes: Ct,
  tablePropertiesRows: ht,
  tablePropertiesColumns: mt,
  tablePropertiesWidth: ft,
  tablePropertiesHeight: Rt,
  tablePropertiesPosition: At,
  tablePropertiesAlignment: ut,
  tablePropertiesCellSpacing: yt,
  tablePropertiesCellPadding: Tt,
  tablePropertiesTableBackground: St,
  tablePropertiesBorderWidth: Bt,
  tablePropertiesBorderColor: It,
  tablePropertiesBorderStyle: wt,
  tablePropertiesCollapseBorders: Ht,
  tablePropertiesId: kt,
  tablePropertiesCssClass: vt,
  tablePropertiesAccessibility: Wt,
  tablePropertiesCaption: Lt,
  tablePropertiesCaptionPosition: Ft,
  tablePropertiesCaptionAlignment: xt,
  tablePropertiesHeaderRows: Mt,
  tablePropertiesHeaderCols: Dt,
  tablePropertiesAssociateHeaders: Nt,
  tablePropertiesGeneral: Ut,
  tablePropertiesAdvanced: Et,
  tablePropertiesSave: Gt,
  tablePropertiesCancel: zt,
  borderStyleNone: Ot,
  borderStyleSolid: Jt,
  borderStyleDotted: Vt,
  borderStyleDashed: jt,
  borderStyleDouble: qt,
  borderStyleGroove: Kt,
  borderStyleRidge: Qt,
  borderStyleInset: Xt,
  borderStyleOutset: Yt,
  borderStyleInitial: Zt,
  borderStyleInherit: _t,
  borderStyleHidden: $t,
  sizeAuto: eo
};
export {
  oo as keys,
  to as messages
};
