// @progress/kendo-react-orgchart/server/ServerOrgChartGroup.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as e from "react";
import { classNames as E, kendoThemeMaps as y } from "@progress/kendo-react-common";
import { ExpandButton as A } from "../client/ExpandButton.mjs";
import { OrgChartGroupContainer as C } from "../client/OrgChartGroupContainer.mjs";
import { GROUP_TITLE_HEIGHT as H, GROUP_SUBTITLE_HEIGHT as L } from "../utils/consts.mjs";
const f = "k-orgchart-node-group", u = "k-orgchart-node-group-title", p = "k-orgchart-node-group-subtitle", x = (o) => {
  const T = {
    orientation: "horizontal"
  }, {
    id: k,
    title: l,
    subtitle: n,
    line: N,
    plus: s,
    focus: U,
    level: t,
    verticalLine: R,
    childLineWidth: c,
    orientation: i,
    expanded: r,
    nodes: a,
    groupTitleHeight: v,
    groupSubtitleHeight: S,
    groupTitleRender: _,
    groupSubtitleRender: G,
    ...O
  } = { ...T, ...o }, m = _, h = G, d = { height: v || H }, g = { height: S || L };
  return /* @__PURE__ */ e.createElement(
    "div",
    {
      ...O,
      className: E(o.className, f, "k-vstack", "k-align-items-center")
    },
    t !== 0 && /* @__PURE__ */ e.createElement(
      "div",
      {
        className: "k-orgchart-line k-orgchart-line-v k-orgchart-line-v-top",
        style: { height: R }
      }
    ),
    /* @__PURE__ */ e.createElement(
      C,
      {
        nodes: a,
        "aria-expanded": !!r,
        "aria-keyshortcuts": "Enter",
        "aria-level": t + 1,
        "aria-owns": k
      },
      m ? /* @__PURE__ */ e.createElement(
        m,
        {
          style: d,
          className: u,
          title: l,
          items: a,
          level: t,
          expanded: r
        }
      ) : /* @__PURE__ */ e.createElement("div", { className: u, style: d }, l),
      h ? /* @__PURE__ */ e.createElement(
        h,
        {
          style: g,
          className: p,
          subtitle: n,
          items: a,
          level: t,
          expanded: r
        }
      ) : /* @__PURE__ */ e.createElement("div", { className: p, style: g }, n),
      /* @__PURE__ */ e.createElement(
        "div",
        {
          role: "group",
          style: { width: "100%" },
          className: E("k-orgchart-node-container", {
            [`k-${y.orientationMap[i] || i}`]: i
          })
        },
        o.children
      )
    ),
    N && /* @__PURE__ */ e.createElement("div", { className: "k-orgchart-line k-orgchart-line-v" }),
    s && /* @__PURE__ */ e.createElement(A, { expanded: r, nodes: a }),
    s && !!c && /* @__PURE__ */ e.createElement("div", { className: "k-orgchart-line k-orgchart-line-h", style: { width: c, marginTop: -15 } })
  );
};
export {
  x as ServerOrgChartGroup
};
