// @progress/kendo-react-orgchart/server/ServerOrgChartNode.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as e from "react";
import { classNames as f } from "@progress/kendo-react-common";
import { Avatar as x, CardTitle as A, CardSubtitle as R } from "@progress/kendo-react-layout";
import { ExpandButton as w } from "../client/ExpandButton.mjs";
import { OrgChartCard as O } from "../client/OrgChartCard.mjs";
import { OrgChartCardBody as T } from "../client/OrgChartCardBody.mjs";
const S = "k-orgchart-node", D = (l) => {
  const h = {
    avatarType: "image"
  }, {
    id: p,
    title: c,
    subtitle: i,
    level: t,
    childLineWidth: m,
    cardWidth: g,
    cardHeight: v,
    verticalLine: E,
    color: k,
    line: N,
    plus: n,
    expanded: a,
    node: r,
    avatar: s,
    avatarType: o,
    itemRender: C,
    showAvatar: u,
    ...y
  } = { ...h, ...l }, d = C;
  return /* @__PURE__ */ e.createElement(
    "div",
    {
      ...y,
      className: f(l.className, S, "k-vstack", "k-align-items-center")
    },
    t !== 0 && /* @__PURE__ */ e.createElement(
      "div",
      {
        className: "k-orgchart-line k-orgchart-line-v k-orgchart-line-v-top",
        style: { height: E }
      }
    ),
    /* @__PURE__ */ e.createElement(
      O,
      {
        node: r,
        style: { width: g, height: v },
        "aria-level": t + 1,
        "aria-expanded": !!a,
        "aria-owns": p
      },
      /* @__PURE__ */ e.createElement(T, { level: t }, d ? /* @__PURE__ */ e.createElement(
        d,
        {
          item: r,
          title: c,
          subtitle: i,
          level: t,
          expanded: a,
          color: k
        }
      ) : /* @__PURE__ */ e.createElement(e.Fragment, null, u && /* @__PURE__ */ e.createElement(x, { type: o, themeColor: "secondary" }, o === "image" ? /* @__PURE__ */ e.createElement("img", { src: s, alt: "KendoReact Avatar Customer" }) : s), /* @__PURE__ */ e.createElement("div", { className: "k-card-title-wrap k-vstack" }, /* @__PURE__ */ e.createElement(A, { className: "k-text-ellipsis" }, c), /* @__PURE__ */ e.createElement("span", { className: "k-spacer" }), /* @__PURE__ */ e.createElement(R, { className: "k-text-ellipsis" }, i)), /* @__PURE__ */ e.createElement("span", { className: "k-spacer" })))
    ),
    N && /* @__PURE__ */ e.createElement("div", { className: "k-orgchart-line k-orgchart-line-v" }),
    n && /* @__PURE__ */ e.createElement(w, { expanded: a, node: r }),
    n && !!m && /* @__PURE__ */ e.createElement("div", { className: "k-orgchart-line k-orgchart-line-h", style: { width: m, marginTop: -15 } })
  );
};
export {
  D as ServerOrgChartNode
};
