// @progress/kendo-react-scheduler/components/ShowMoreItemsButton.mjs
/**
 * @license
 *-------------------------------------------------------------------------------------------
 * Copyright © 2025 Progress Software Corporation. All rights reserved.
 * Licensed under commercial license. See LICENSE.md in the package root for more information
 *-------------------------------------------------------------------------------------------
 */
import * as e from "react";
import { Button as s } from "@progress/kendo-react-buttons";
import { useLocalization as u } from "@progress/kendo-react-intl";
import { moreEvents as r, messages as f } from "../messages/index.mjs";
import { moreHorizontalIcon as d } from "@progress/kendo-svg-icons";
const k = e.forwardRef(
  (t, a) => {
    const o = e.useRef(null), n = e.useRef(null), l = u();
    e.useImperativeHandle(o, () => ({
      element: n.current && n.current.element,
      ...t
    })), e.useImperativeHandle(a, () => o.current);
    const c = e.useCallback(
      (i) => {
        !t.onClick || !o.current || t.onClick.call(void 0, {
          target: o.current,
          syntheticEvent: i
        });
      },
      [o, t.slot, t.onClick]
    ), m = l.toLanguageString(r, f[r]);
    return /* @__PURE__ */ e.createElement(
      s,
      {
        ref: n,
        tabIndex: -1,
        className: "k-more-events",
        onClick: c,
        "aria-label": m,
        icon: "more-horizontal",
        svgIcon: d
      }
    );
  }
);
k.displayName = "KendoReactSchedulerShowMoreItemsButton";
export {
  k as ShowMoreItemsButton
};
